core/src/main/java/hudson/model/DirectoryBrowserSupport.java

```diff
@@ -375,6 +378,14 @@ private void serveFile(StaplerRequest req, StaplerResponse rsp, VirtualFile root
         if (LOGGER.isLoggable(Level.FINE))
             LOGGER.fine("Serving " + baseFile + " with lastModified=" + lastModified + ", length=" + length);
 
+        if (!SystemProperties.getBoolean(DISABLE_ARTIFACT_DEFAULT_CHARSET, false)) {
+            String fileEncoding = System.getProperty("file.encoding");
+            if (fileEncoding != null) {
+                // JENKINS-28676 serve artifacts using the default encoding of the JVM
+                rsp.setContentType("text/plain;charset=\"" + fileEncoding + "\"");
```

Comment:
This makes the comment in line 401 obsolete.

Reply:
Also, looks like https://github.com/jenkinsci/stapler/blob/3aa95fe179cb5534806b05d2f41947791a78b928/core/src/main/java/org/kohsuke/stapler/Stapler.java#L662-L663 via https://github.com/jenkinsci/stapler/blob/3aa95fe179cb5534806b05d2f41947791a78b928/core/src/main/java/org/kohsuke/stapler/Stapler.java#L571-L572 makes https://github.com/jenkinsci/stapler/blob/3aa95fe179cb5534806b05d2f41947791a78b928/core/src/main/java/org/kohsuke/stapler/Stapler.java#L132-L140 the canonical way to do this.